### PR TITLE
[Merged by Bors] - feat: make push_neg fail if it didn't do anything (like simp)

### PR DESCRIPTION
**Pull request by alexjbest**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Tactic/PushNeg.lean [hunk 1]
```diff
@@ -181,7 +181,10 @@ macro (name := pushNeg) tk:"#push_neg " e:term : command => `(command| #conv%$tk
 def pushNegTarget : TacticM Unit := withMainContext do
   let goal ← getMainGoal
   let tgt ← instantiateMVars (← goal.getType)
-  replaceMainGoal [← applySimpResultToTarget goal tgt (← pushNegCore tgt)]
+  let newGoal ← applySimpResultToTarget goal tgt (← pushNegCore tgt)
+  if newGoal == goal then throwError "push neg made no progress"
```
Comment (adomani) [inline Mathlib/Tactic/PushNeg.lean:185]:
```suggestion
  if newGoal == goal then throwError "push_neg made no progress"
```
maybe?

File: Mathlib/Tactic/PushNeg.lean [hunk 2]
```diff
@@ -191,6 +194,7 @@ def pushNegLocalDecl (fvarId : FVarId) : TacticM Unit := withMainContext do
   let goal ← getMainGoal
   let myres ← pushNegCore tgt
   let some (_, newGoal) ← applySimpResultToLocalDecl goal fvarId myres False | failure
+  if newGoal == goal then throwError "push neg made no progress"
```
Comment (adomani) [inline Mathlib/Tactic/PushNeg.lean:197]:
```suggestion
  if newGoal == goal then throwError "push_neg made no progress"
```
like above.

**Review by adomani — COMMENTED** [review]
I like this!

If I manage to speed up the unused tactic linter, these uses would be flagged by it as well.

**Comment by j-loreaux** [comment]
excellent, thanks!

bors merge

**Pull request closed without merging**